### ISSUE TITLE
Output shape of dilated convolution, I suspect it is a bug 

### ISSUE THREAD

**Issue by lood339** [vlfeat/matconvnet#1010]
In the vl_nnconv.m, the document says
%   The output a is an array of dimension YH x YW x K x N of N images
%   with K feature challens and size:
%
%     YH = floor((H + (PADTOP+PADBOTTOM) - FH)/STRIDEY) + 1,
%     YW = floor((W + (PADLEFT+PADRIGHT) - FW)/STRIDEX) + 1. This is right.
%
%   Accounting for dilation, the formulas become:
%
%     YH = floor((H + (PADTOP+PADBOTTOM) - FH*(DILATEY-1) -1)/STRIDEY) + 1,
%     YW = floor((W + (PADLEFT+PADRIGHT) - FW*(DILATEX-1) -1)/STRIDEX) + 1. This might be a bug.

For the dilation, when DILATEY = 1, the output shape (YH) is not related to filter size (FH). I cross check with caffe implementation.
This formulation is slightly different from the implementation of caffe.
_const int input_dim = this->input_shape(i + 1);
    const int kernel_extent = dilation_data[i] * (kernel_shape_data[i] - 1) + 1;  // actual kernel size
    const int output_dim = (input_dim + 2 * pad_data[i] - kernel_extent)
/ stride_data[i] + 1;

I think this implementation is more straightforward. The dilation increases the size of kernel_extent. When the kernel size is 3, dilation is 2 and the input size 6, stride is 1 and no padding. The output put shape should be 2, but the MatconvNet output shape is 3.

**Comment by ZhaMaoGe** [comment]
I have the same question, maybe the document have been wrong?

**Comment by QiaoLiuHit** [comment]
Yes, the document on the website is wrong but the implementation of the code is right.
the right form is YW=(W+P_left+P_right-D*(F-1)-1)/S+1.

**Comment by Tveek** [comment]
output_size=w+2*pad-(d(k-1)+1),this is more clear.

**Comment by phdsky** [comment]
> output_size=(w+2*pad-(d(k-1)+1))/s+1,this is more clear.

Great! Matching the code in caffe:

```
template <typename Dtype>
void ConvolutionLayer<Dtype>::compute_output_shape() {
  const int* kernel_shape_data = this->kernel_shape_.cpu_data();
  const int* stride_data = this->stride_.cpu_data();
  const int* pad_data = this->pad_.cpu_data();
  const int* dilation_data = this->dilation_.cpu_data();
  this->output_shape_.clear();
  for (int i = 0; i < this->num_spatial_axes_; ++i) {
    // i + 1 to skip channel axis
    const int input_dim = this->input_shape(i + 1);
    const int kernel_extent = dilation_data[i] * (kernel_shape_data[i] - 1) + 1;
    const int output_dim = (input_dim + 2 * pad_data[i] - kernel_extent)
        / stride_data[i] + 1;
    this->output_shape_.push_back(output_dim);
  }
}
```
